pysindy/optimizers/stlsq.py

```diff
@@ -125,28 +130,54 @@ def __init__(
         self.ridge_kw = ridge_kw
         self.initial_guess = initial_guess
         self.verbose = verbose
+        self.sparse_ind = sparse_ind
 
-    def _sparse_coefficients(self, dim, ind, coef, threshold):
+    def _sparse_coefficients(self, dim, ind_nonzero, coef, threshold):
         """Perform thresholding of the weight vector(s)"""
         c = np.zeros(dim)
-        c[ind] = coef
+        c[ind_nonzero] = coef
         big_ind = np.abs(c) >= threshold
-        c[~big_ind] = 0
+        if self.sparse_ind is not None:
+            sparse_ind_mask = np.zeros_like(ind_nonzero)
+            sparse_ind_mask[self.sparse_ind] = True
+            c[~big_ind & sparse_ind_mask] = 0
+        if self.sparse_ind is None:
+            c[~big_ind] = 0
         return c, big_ind
 
-    def _regress(self, x, y):
+    def _regress(self, x, y, dim, sparse_sub):
         """Perform the ridge regression"""
         kw = self.ridge_kw or {}
-
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", category=LinAlgWarning)
-            try:
-                coef = ridge_regression(x, y, self.alpha, **kw)
-            except LinAlgWarning:
-                # increase alpha until warning stops
-                self.alpha = 2 * self.alpha
-        self.iters += 1
-        return coef
+        sparse_ind = self.sparse_ind
+        if sparse_ind is None:
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=LinAlgWarning)
+                try:
+                    coef = ridge_regression(x, y, self.alpha, **kw)
+                except LinAlgWarning:
+                    # increase alpha until warning stops
+                    self.alpha = 2 * self.alpha
+            self.iters += 1
+            return coef
+        if sparse_ind is not None:
+            coef = np.zeros(dim)
```

Comment:
I don't believe this line is necessary.  `try...except` behaves differently with `Warning` objects.  Since `Warnings` do not cause execution to break, the entire `try` block will execute, and then the `except` block will execute.  This means that `coef` will always be assigned a value in the `try...except`, and you can delete `coef = np.zeros(dim)`.